### DISCUSSION TITLE
Script component

### DISCUSSION THREAD

**Discussion by TpTTT-T** [vercel/next.js#36945] [Help]
> The reason beforeInteractive was designed to work only inside \_document.js is to support streaming and Suspense functionality. Outside of the _document, it's not possible to guarantee the timing or ordering of beforeInteractive scripts.

This is doc about [beforeInteractive](https://nextjs.org/docs/basic-features/script#beforeinteractive)
The doc descride the reason that beforeInteractive only work insidex _document.js, but I still do not understand what is steaming and suspense meaining.

**Comment by icyJoseph** [comment]
It is talking about a new React feature, which allows server side rendering to be done in parts.

Normally, server side rendering was a synchronous process which resulted in a string, also know as dehydrated HTML. This is what the server sent to the client, so that it could be hydrated.

There's new developments in React, which will make it so that the server can render pieces of HTML and send them to the client (this is an extreme simplification). The `_document` head elements (scripts and such, that need to be present before interactivity) as far as I understand, will be streamed to the client in one go, so that those resources can be fetched by the browser ASAP.

Here you can read way more about it:

- https://github.com/reactwg/react-18/discussions/37
- https://github.com/reactwg/react-18/discussions/47#discussioncomment-847004

I'd say though, don't lose sleep over it. When it's fully ready and the community understands the ins and outs, you can pour more time into it.
